### ISSUE TITLE
Error while restoring specific databases: pg_filenode.map" contains invalid data

### ISSUE THREAD

**Issue by devrimgunduz** [pgbackrest/pgbackrest#986]
1. pgBackRest version: 
2.25

2. PostgreSQL version: 
11.7 

3. Operating system/version :
CentOS 7.7, all updates applied

4. Did you install pgBackRest from source or from a package? 
Yes: PGDG RPMs

5. Describe the issue:
Tried to restore two databases from a backup with:

pgbackrest --stanza=db-primary restore --reset-pg1-host --db-include=bar1 --db-include=foo1 --pg1-path=/backup/pgrestore/  

postmaster starts fine, but even after replaying WAL logs, I got this:

-bash-4.2$ psql
psql: error: could not connect to server: FATAL:  relation mapping file "base/18778/pg_filenode.map" contains invalid data

(FWIW, this is the "postgres" database)

So, even though the docs say that postgres, template1 and template0 are always restored, there is something wrong here. To verify, I added postgres to the  list of included databases, and then I got:

-bash-4.2$ /usr/pgsql-11/bin/psql 
psql (11.7)
Type "help" for help.

So, back to the  error message: the OID of postgres is 18778, not something < 16384. I suspect that the code may be checking OIDs < 16384 while assuming that it will always restore postgres, template0 and template1.

Not sure whether to classify this is as a code bug, or documentation issue. Apparently pgbackrest will restore postgres database if and only if the OID < 16384.

Comments?

**Comment by sfrost** [comment]
Considering that in the common case today (where the postgres database *hasn't* been dropped/recreated), we will end up restoring it (right?), it seems like we should make the code actually reflect the documentation and always include the 'postgres' database in what we restore- but then, I feel like we then need a 'db-exclude' option, as some folks might have pretty large 'postgres' databases that they'd like to have excluded from being restored.

We already collect up the database->OID mapping during the backup, don't we..?  Is there a reason why we need to base the databases on the 16384 number?  I mean, users can drop/recreate template0 and template1 too, if they really wanted to....

**Comment by woblerr** [comment]
We get the same problem for one of our instances.

* pgBackRest version: 2.30
* PostgreSQL version: 11.9

We restore some specific database `foo` from last backup and replaying WAL by command like:
```bash
pgbackrest restore --stanza foo-db --db-include foo --type time --target "2021-02-17 07:00:00" --target-action=promote  --tablespace-map-all /var/lib/pgsql/11/tablespace --log-level-console info
```

If we try to connect to `postgres` database, we get an error: `psql: FATAL:  relation mapping file "base/17042/pg_filenode.map" contains invalid data`.
If we connect to `foo` db everything is ok.

After specify `postgres` in `restore` command everythins is ok. Command like:
```bash
pgbackrest restore --stanza foo-db --db-include foo --db-include postgres --type time --target "2021-02-17 07:00:00" --target-action=promote --tablespace-map-all /var/lib/pgsql/11/tablespace --log-level-console info
```

When we checked OID, we found that OID system db (`postgres (17042)`) more than [16384](https://github.com/pgbackrest/pgbackrest/blob/d29855bd0bf8b957fa8c06445d22e425c443a704/src/postgres/interface.h#L75):  [restore.c#L1252](https://github.com/pgbackrest/pgbackrest/blob/d29855bd0bf8b957fa8c06445d22e425c443a704/src/command/restore/restore.c#L1252) and `postgres` db was restored as zero db.

```bash
pgbackrest info --stanza foo-db --set 20210214-010010F_20210217-010009D

stanza: foo-db
    status: ok
    cipher: none

    db (current)
        wal archive min/max (11-1): 000000010000003100000063/000000010000003C000000F5

        diff backup: 20210214-010010F_20210217-010009D
            timestamp start/stop: 2021-02-17 01:00:09 / 2021-02-17 01:05:58
            wal start/stop: 000000010000003C000000C9 / 000000010000003C000000C9
            database size: 16GB, backup size: 8.8GB
            repository size: 2.7GB, repository backup size: 1.6GB
            backup reference list: 20210214-010010F
            database list: foo (16402), bar (16796), postgres (17042), template1 (16401)
            tablespaces:
                sample (16400) => /pg_tablespaces/bar
```

@dwsteele any ideas?

**Comment by pgstef** [comment]
The [16384](https://github.com/postgres/postgres/blob/e6b8e83b9f012cfbb3a2f96a7d2e74aebd299d4e/src/include/access/transam.h#L156) hasn't changed in PG itself. I guess that you have an OID higher than that because you re-created the pg database manually ?

If yes, we could (A) adjust the documentation to explicitly tell to include the system (template0, template1, postgres) databases recreated manually. Or (B) like Stephen suggested, figure out the ids of those dbs from the backup manifest and exclude them.

**Comment by woblerr** [comment]
@pgstef unfortunately, I don't have the exact information about manual re-creation postgres db, but it looks like this. So I will also assume that the OID has changed due to the re-creation and the answer is yes.

A) It seems that this is the minimum action required. For example, the question about including manually recreated system databases can be described in the FAQ section.

B) When restoring with the `--db-include` option, both options that @sfrost suggested are possible:
* using an additional option, for example, `--db-exclude` in case there is no need to restore postgres database.
In this case user can exclude postgres database even if OID less then `16384`.
* not restoring system databases automatically (at least postgresql).
But when you need to allow them to be restored using the `--db-include` option.

Logically, if the user consciously specifies which databases he wants to restore, then it makes no sense to restore the system databases, if there are no technical restrictions.

**Comment by pgstef** [comment]
I've got a [poc](https://github.com/pgstef/pgbackrest/tree/issues/986) around the B option.

Let's start with a backup where the user re-creates `postgres`. The id changes and it is reported as "user database" even in the `info` command:
```
        full backup: 20210218-130153F
            timestamp start/stop: 2021-02-18 13:01:53 / 2021-02-18 13:02:05
            wal start/stop: 00000001000000000000001E / 00000001000000000000001E
            database size: 487.3MB, database backup size: 487.3MB
            repo1: backup set size: 29MB, backup size: 29MB
            database list: postgres (16386), test (16384), test2 (16385)
```

When using `db-include`, it will fetch `template0`, `template1` and `postgres` oid from the backup manifest to avoid erasing it on restore by default. Log example with `--db-include=test`:
```
DETAIL: databases found for selective restore (1, 13318, 16384, 16385, 16386)
DETAIL: system database id to include : 13318 (template0)
DETAIL: system database id to include : 1 (template1)
DETAIL: system database id to include : 16386 (postgres)
DETAIL: database with id 16385 will be zeroed
```

Now, if we want to restore all the databases without that test db, using `--db-exclude=test`:
```
DETAIL: databases found for selective restore (1, 13318, 16384, 16385, 16386)
DETAIL: system database id to include : 13318 (template0)
DETAIL: system database id to include : 1 (template1)
DETAIL: system database id to include : 16386 (postgres)
DETAIL: database with id 16384 will be zeroed
´´´

And if we only want test without postgres, using `--db-include=test --db-exclude=postgres` together:
```
DETAIL: databases found for selective restore (1, 13318, 16384, 16385, 16386)
DETAIL: system database id to include : 13318 (template0)
DETAIL: system database id to include : 1 (template1)
DETAIL: system database id to include : 16386 (postgres)
DETAIL: database with id 16385 will be zeroed
DETAIL: database with id 16386 will be zeroed
```

Does it sounds nice from a user point of view? If so, I could work on the coverage tests to submit the PR.

**Comment by woblerr** [comment]
In the case when postgres db was not recreated (OID < `16384`), there is no possibility to exclude it using the flag `--db-exclude`.
Just like it works now. Right?

Look good to me. Thanks a lot!

**Comment by pgstef** [comment]
> In the case when `postgres` db was not re-created (OID < `16384`), there is no possibility to exclude it using the flag `--db-exclude`.
> Just like it works now. Right?

Indeed, the current [poc](https://github.com/pgstef/pgbackrest/blob/22320503cfc732af4ce2706dc9ecfc496e36eb59/src/command/restore/restore.c#L1300) doesn't allow "system" tables to be zeroed, but that's just a condition to change if we'd like to extend that possibility.

**Comment by woblerr** [comment]
Probably, the optimal behavior should be like this:

When restoring using the `--db-include` flag, the system databases (postgres, template0, template1) are also restored by default,
even if they were re-created manually.

Using the `--db-exclude` flag for `restore` command, you can exclude restoring system databases regardless of their OIDs.
But at the same time, an additional warning  is needed to display that one or more system databases will be restored to zero.

**Comment by cmwshang** [comment]
I agree with @woblerr. My concern is that it appears the root of the problem is that the assumption that the system databases never have an OID > 16384 is clearly not always correct.

**Comment by cmwshang** [comment]
Also, I am curious if there is any connection between this issue and https://github.com/pgbackrest/pgbackrest/issues/796 where `recovery_target = 'immediate'` appeared to help.

**Comment by pgstef** [comment]
Indeed, excluding any db (even a `postgres` not recreated by the user) could be a very nice feature.
I've reworked the patch and added the tests to submit the PR. 

The automatic include won't probably fix the underlying problem but provide a quick solution to users I guess.

**Comment by dwsteele** [comment]
Closing since this issue appears to be addressed by PR #1307. Anyone who is interested should review and comment there.
